### PR TITLE
Whirlwind support for Barbarians. 

### PR DESCRIPTION
**Pull request by jayRyan24**
Added support for whirlwind.

### CODE REVIEW

**Comment by jayRyan24** [comment]
Example:

https://pastebin.com/nZY72MGU

**Comment by jayRyan24** [comment]
Also, I've been running it with all my chars without an issue. 

Warrior Mage:
`A black ape roars and strains against the web that binds it.
[whirlwind-trainer]>attack
Your foot is engulfed by fire as bright as the noontime sun!
< Moving as a single sinuous force, you kick your foot at a black ape.  A black ape fails to dodge, only
  partially avoiding the impact.  The foot lands a solid hit (4/23) to the ape's chest.
[You're adeptly balanced and overwhelming opponent.]
[Roundtime 3 sec.]
[whirlwind-trainer]>pre MOA 1
>
You trace a careful sigil in the air, shaping the pattern of the Mark of Arhat spell.
[whirlwind-trainer]>cast
You gesture at a black ape.
A dancing flame springs into existence between you and a black ape, quickly growing into a raging inferno.  A
  large golden lion steps proudly out of the fire, drawing the flames with it like a shroud.  Baring brightly
  glowing fangs, the lion pounces on the black ape!

The lion slashes at the black ape with its blazing claws, leaving a trail of jagged yellow-white lines branded
  upon it!
The fiery lion shakes its glowing mane, then vanishes in a swirl of flames.
The complementary nature of the spell empowers you.
Roundtime: 2 sec.`

Necro:
`[whirlwind-trainer]>gouge
You leap out of hiding and ambush a snowbeast, catching it by surprise.
< With the speed and temerity of a swooping falcon, you gouge your fingers at a snowbeast.  A snowbeast fails to
  evade, the blow brushing it.  The fingers lands a light hit (1/23) to the snowbeast's left hand.
[You're incredibly balanced and overwhelming opponent.]
[Roundtime 1 sec.]
>
* Ineptly, a snowbeast claws at you.  You dodge, twisting to one side.
[You're incredibly balanced and overwhelming your opponent.]
[whirlwind-trainer]>cast
You gesture at a snowbeast.
You whisper several words of arcane power, attempting to twist a snowbeast's mind with primal terror.
Your malefic assault overtakes a snowbeast, petrifying it to the spot!
Roundtime: 2 sec.`

Ranger:
`[whirlwind-trainer]>cast
You gesture at a fledgling forest gryphon.
You focus on the forest gryphon's life essence as you swipe your hand downward in its direction!
A bluish green aura leaches out from the forest gryphon, which solidifies into a double helix composed of
  the two individual hues.  The helices spin rapidly as it shrinks into the forest gryphon's chest.
Roundtime: 2 sec.
>
You feel fully rested.
[whirlwind-trainer]>rub my Everild orb
The Everild orb feels warm to your touch and has a strong, steady glow.  The strange tugging sensation is
  gone, leading you to believe that your sacrifice is properly prepared.
[whirlwind-trainer]>hide
>
You melt into the background, convinced that your attempt to hide went unobserved.
Roundtime: 2 sec.
>
A baby forest gryphon fans its wings, sniffing the air alertly.
[whirlwind-trainer]>jab
You leap out of hiding and ambush a fledgling forest gryphon, startling it.
< Moving with grace, you jab a steel cutlass at a fledgling forest gryphon.  A fledgling forest gryphon
  fails to dodge.  The cutlass lands a light hit (1/23) to the gryphon's right arm.
[You're adeptly balanced and in superior position.]
[Roundtime 2 sec.]`

File: combat-trainer.lic [hunk 1]
```diff
@@ -171,13 +176,27 @@ class SetupProcess
...
     # Invoke Focus
     return unless game_state.weapon_skill == 'Targeted Magic' && game_state.weapon_name
     bput("invoke #{game_state.weapon_name}", 'You')
     waitrt?
   end
+
+  def determine_whirlwind_weapon(game_state)
+    temp_array = (game_state.whirlwind_trainables - [game_state.weapon_skill])
+    offhand_skill = temp_array.min_by { |skill| [DRSkill.getxp(skill), DRSkill.getrank(skill)] }
+    game_state.update_whirlwind_weapon_info(offhand_skill)
```
Comment (rpherbig) [inline combat-trainer.lic]: Does this account for multiple weapon skills using the same weapon (e.g. swappable weapons)?

File: combat-trainer.lic [hunk 2]
```diff
@@ -1844,6 +1874,7 @@ class GameState
     @aim_queue = []
     @dance_queue = []
     @analyze_combo_array = []
+    @currently_whirlwinding = false
```
Comment (rpherbig) [inline combat-trainer.lic]: Can you add these new vars to the declarations? In my copy that's lines 1812, 2326, or 2327.

File: drinfomon.lic [hunk 3]
```diff
@@ -753,9 +753,9 @@ while line = script.gets
         end
     elsif line =~ /^Circle: (\d+)/
       DRStats.circle = Regexp.last_match(1).to_i
-    elsif line =~ /^\[You're(.*\,)?(.*?)(and|with.*)?\]$/
-      balance_map = ['completely imbalanced', 'hopelessly unbalanced', 'extremely imbalanced', 'very badly balanced', 'badly balanced', 'somewhat off balance', 'off balance', 'slightly off balance', 'solidly balanced', 'nimbly balanced', 'adeptly balanced', 'incredibly balanced']
-      DRStats.balance = balance_map.index(Regexp.last_match(2).strip)
+    elsif line =~ /(?<=You're\s)(?<balance>\w*\s*\w+)(?=\s\w*(balance))/
```
Comment (rpherbig) [inline drinfomon.lic]: This would be a great PR on its own :)

File: profiles/Chuno-stealth.yaml [hunk 4]
```diff
@@ -0,0 +1,31 @@
+---
```
Comment (rpherbig) [inline profiles/Chuno-stealth.yaml]: All of these YAML updates (minus the whirlwind specific changes) would be a good separate PR

File: combat-trainer.lic [hunk 5]
```diff
@@ -1940,11 +1974,14 @@ class GameState
     @attack_overrides = settings.attack_overrides
     echo("  @attack_overrides: #{@attack_overrides}") if $debug_mode_ct
 
-    # For deprecation support
```
Comment (rpherbig) [inline combat-trainer.lic]: This would be a good separate PR

File: combat-trainer.lic [hunk 6]
```diff
@@ -171,13 +176,27 @@ class SetupProcess
       game_state.prepare_summoned_weapon(last_summoned)
     else
       @equipment_manager.wield_weapon(game_state.weapon_name, game_state.weapon_skill)
+      if game_state.use_whirlwind? && game_state.is_whirlwind_trainable?(game_state.weapon_skill)
```
Comment (rpherbig) [inline combat-trainer.lic]: If `use_whirlwind` is only used in one place, it's a reasonable candidate for pulling out of GameState and making local to the process. This is not something we've been very consistent on... but GameState is pretty big already...

File: combat-trainer.lic (hunk 1 above)
Comment (jayRyan24) [inline combat-trainer.lic]: Nah, I'm not sure how you would implement that. I'll add non-swappable to the validate.

Comment (rpherbig) [inline combat-trainer.lic]: Could you do something like `.reject { |skill| weapon_training[skill] == game_state.weapon_name }` ?

File: combat-trainer.lic [hunk 7]
```diff
@@ -2133,6 +2218,12 @@ class GameState
     DRStats.fatigue < @fatigue_regen_threshold
   end
 
+  def balance_low?
+    echo("***Balance: #{DRStats.balance}***") if $debug_mode_ct
+    echo("***Target: #{@balance_regen_threshold}***") if $debug_mode_ct
+    DRStats.balance < @balance_regen_threshold
```
Comment (rpherbig) [inline combat-trainer.lic]: Is the balance threshold change necessary for whirlwind, or could it be a separate PR?

File: combat-trainer.lic [hunk 8]
```diff
@@ -171,13 +176,27 @@ class SetupProcess
       game_state.prepare_summoned_weapon(last_summoned)
     else
       @equipment_manager.wield_weapon(game_state.weapon_name, game_state.weapon_skill)
+      if game_state.whirlwind_trainables && game_state.is_whirlwind_trainable?(game_state.weapon_skill)
```
Comment (rpherbig) [inline combat-trainer.lic]: What do you think about having `is_whirlwind_trainable?` return false if `whirlwind_trainables` is empty? Then you only need the second clause in this `if`.

File: combat-trainer.lic [hunk 9]
```diff
@@ -450,9 +475,14 @@ class LootProcess
       @last_rites_timer = Time.now
       return
     end
+
+    @equipment_manager.stow_weapon(game_state.whirlwind_offhand_name) if game_state.currently_whirlwinding?
```
Comment (rpherbig) [inline combat-trainer.lic]: GameState has a copy of equipment manager. I suspect this line could be moved there and be a bit simpler. What do you think?

Comment (rpherbig) [inline combat-trainer.lic]: And the same with line 484

**Comment by jayRyan24** [comment]
I can't figure out what the the world the issue is with this conflict?

File: combat-trainer.lic (hunk 8 above)
Comment (jayRyan24) [inline combat-trainer.lic]: Good call, added this.

**Comment by jayRyan24** [comment]
Oh yeah, my other PR's back merges. Bleh.

File: combat-trainer.lic [hunk 10]
```diff
@@ -1999,6 +2036,62 @@ class GameState
...
+  end
+
+  def determine_whirlwind_action
+    if @use_weak_attacks && Flags["ct-accuracy-ready"]
+      update_analyze_array(true) if @analyze_combo_array.empty?
+      return @analyze_combo_array.shift
+    elsif (npcs.length > 2) && !balance_low? 
+      return 'whirlwind' 
+    elsif balance_low? 
+      return 'gouge'
```
Comment (rpherbig) [inline combat-trainer.lic]: You can `gouge` even if both hands have weapons?

File: combat-trainer.lic [hunk 11]
```diff
@@ -1999,6 +2036,62 @@ class GameState
...
+
+  def whirlwind_trainables
+    @whirlwind_trainables
+  end
+
+  def whirlwind_offhand_name
+    weapon_training[@current_whirlwind_offhand_skill]
+  end
+
+  def last_whirlwind_offhand_weapon_name
```
Comment (rpherbig) [inline combat-trainer.lic]: Where is this method used?

File: combat-trainer.lic [hunk 12]
```diff
@@ -1999,6 +2036,62 @@ class GameState
...
+    @last_whirlwind_offhand_skill = @current_whirlwind_offhand_skill
+    @current_whirlwind_offhand_skill = weapon_skill
+  end
+
+  def whirlwind_offhand_skill
+    @current_whirlwind_offhand_skill
+  end
+
+  def is_whirlwind_trainable?(weapon_skill)
+    return false if !whirlwind_trainables.any?
```
Comment (rpherbig) [inline combat-trainer.lic]: This could be `return false if whirlwind_trainables.empty?`

File: profiles/base.yaml [hunk 13]
```diff
@@ -54,6 +54,7 @@ charged_maneuvers:
   Twohanded Blunt: crash
   Staves: twirl
   Polearms: impale
+  Dual Wield: doublestrike
```
Comment (rpherbig) [inline profiles/base.yaml]: You can't see it, but I'm doing a happy dance. This will close #9!

File: profiles/base.yaml [hunk 14]
```diff
@@ -93,6 +95,7 @@ use_weak_attacks: false
 loot_bodies: true
 dump_junk: false
 use_barb_combos: false
+whirlwind_trainables:
```
Comment (rpherbig) [inline profiles/base.yaml]: Can you add an entry to `base-empty.yaml` for this? I think it would be `[]`.

File: combat-trainer.lic [hunk 15]
```diff
@@ -1999,6 +2036,62 @@ class GameState
...
+  def update_whirlwind_weapon_info(weapon_skill)
+    @last_whirlwind_offhand_skill = @current_whirlwind_offhand_skill
+    @current_whirlwind_offhand_skill = weapon_skill
+  end
+
+  def whirlwind_offhand_skill
+    @current_whirlwind_offhand_skill
+  end
+
+  def is_whirlwind_trainable?(weapon_skill)
```
Comment (rpherbig) [inline combat-trainer.lic]: If the input is _always_ `game_state.weapon_skill`, what do you think about using GameState's instance var instead of a function parameter?

File: combat-trainer.lic (hunk 1 above)
Comment (rpherbig) [inline combat-trainer.lic]: Sounds good

File: combat-trainer.lic [hunk 16]
```diff
@@ -1999,6 +2036,62 @@ class GameState
     @attack_overrides[weapon_skill]
   end
 
+  def set_currently_whirlwinding(setting)
```
Comment (rpherbig) [inline combat-trainer.lic]: Since you defined `currently_whirlwinding` in the attr_accessor list, you don't need this set function. You can `game_state.currently_whirlwinding = true` in a process.

File: combat-trainer.lic [hunk 17]
```diff
@@ -1999,6 +2036,62 @@ class GameState
...
+    @currently_whirlwinding = setting
+  end
+
+  def currently_whirlwinding?
```
Comment (rpherbig) [inline combat-trainer.lic]: Likewise, this method isn't necessary - the only different would be you wouldn't have a `?` if you used the one provided by `attr_accessor`

File: combat-trainer.lic [hunk 18]
```diff
@@ -1999,6 +2036,62 @@ class GameState
...
+    @currently_whirlwinding
+  end
+
+  def whirlwind_trainables
```
Comment (rpherbig) [inline combat-trainer.lic]: Likewise, this method isn't necessary

File: combat-trainer.lic [hunk 19]
```diff
@@ -1895,6 +1926,9 @@ class GameState
     @fatigue_regen_threshold = settings.fatigue_regen_threshold
     echo("  @fatigue_regen_threshold: #{@fatigue_regen_threshold}") if $debug_mode_ct
 
+    @balance_regen_threshold = settings.balance_regen_threshold
```
Comment (rpherbig) [inline combat-trainer.lic]: Can you add this to the declarations? Since it's 'read-only', it should probably be in the `attr_reader` list (which only generates a 'getter' function) on line 2420.

File: combat-trainer.lic (hunk 10 above)
Comment (jayRyan24) [inline combat-trainer.lic]: Yes, otherwise I wouldn't have it in there. :P

File: combat-trainer.lic (hunk 11 above)
Comment (jayRyan24) [inline combat-trainer.lic]: It's not used, but thought to input it as we have it for last_weapon, etc. Can be removed without an issue.

File: combat-trainer.lic (hunk 12 above)
Comment (jayRyan24) [inline combat-trainer.lic]: Yeah, this bad logic is just because I quick fixed something bad, heh. Changed.

File: combat-trainer.lic (hunk 10 above)
Comment (jayRyan24) [inline combat-trainer.lic]:
```
[whirlwind-trainer]>whirlwind
You swing your war hammer with wild abandon, weaving it through the air in a flurry of blows that are nearl
  impossible to count.  You channel the momentum with surreal skill and grace.

Your war hammer lands a good strike (3/23) to a young wyvern's muscular right !
The young wyvern is lightly stunned!

Your throwing axe lands a good strike (3/23) to a young wyvern's scaly chest!

Your war hammer lands a good hit (2/23) to a young wyvern's right hind claw!

A young wyvern deftly sidesteps your throwing axe.

Your war hammer lands a good hit (2/23) to a young wyvern's scaly chest!

A young wyvern deftly sidesteps your throwing axe.

[You're somewhat off balance]
Roundtime: 4 sec.
>
Cutter Giulietta just arrived.
>
* Moving weakly, a young wyvern slashes its razor-sharp hind talons at you.  You evade, barely leaning out
  of harm's reach.
[You're winded, somewhat off balance and opponent has slight advantage.]
>
* Moving poorly, a young wyvern slashes its razor-sharp hind talons at you.  You evade, leaning to one side
  with no room to spare.
[You're winded, somewhat off balance with opponent in good position.]
[whirlwind-trainer]>berserk wildfire
But you are already enraged with that berserk.
[whirlwind-trainer]>berserk earthquake
The young wyvern closes to pole weapon range on you!
[whirlwind-trainer]>gouge
>
Giulietta gets a tyrium serrated parazonium from inside her rapier sheath.
>
But you are already enraged with that berserk.
>
< Moving well, you gouge your fingers at a young wyvern.  A young wyvern fails to evade.  The fingers lands
  a light hit (1/23) to the wyvern's left arm.
[You're winded, badly balanced with opponent in better position.]
[Roundtime 1 sec.]
````

Comment (rpherbig) [inline combat-trainer.lic]: Neat!

File: profiles/base.yaml (hunk 13 above)
Comment (jayRyan24) [inline profiles/base.yaml]: 🗡

File: profiles/base.yaml (hunk 14 above)
Comment (jayRyan24) [inline profiles/base.yaml]: Done

File: combat-trainer.lic (hunk 16 above)
Comment (jayRyan24) [inline combat-trainer.lic]: Done

File: combat-trainer.lic (hunk 17 above)
Comment (jayRyan24) [inline combat-trainer.lic]: Done

File: combat-trainer.lic (hunk 18 above)
Comment (jayRyan24) [inline combat-trainer.lic]: Done.

File: combat-trainer.lic (hunk 19 above)
Comment (jayRyan24) [inline combat-trainer.lic]: Ah, so that's what those things are for. This is very informative, and shows how lazy I am. Thanks. Done.

File: combat-trainer.lic (hunk 15 above)
Comment (jayRyan24) [inline combat-trainer.lic]: Think that's probably the best use.

File: combat-trainer.lic (hunk 10 above)
Comment (jayRyan24) [inline combat-trainer.lic]: Yeah, learned that from Cainec. All of these things that put me off from getting this setup, turned out to be things that weren't very difficult to handle or non issues.

File: equipmanager.lic [hunk 20]
```diff
@@ -151,6 +151,18 @@ class EquipmentManager
     end
   end
 
+  def wield_weapon_offhand(description)
```
Comment (rpherbig) [inline equipmanager.lic]: I noticed that `wield_weapon` has a special case for the Offhand skill. Is that a responsibility of the equipment manager, or should combat-trainer be figuring that out and using one of these two `wield*` functions? This isn't part of the PR, but maybe food for thought...

File: validate.lic [hunk 21]
```diff
@@ -277,6 +277,46 @@ class DRYamlValidator
             .each { |skill, item| error("Item in weapon_training could not be found in gear listings. '#{skill}: #{item}'") }
   end
 
+  def assert_minimum_whirlwind_trainables(settings)
+    return unless settings.whirlwind_trainables
+    return unless setting.whirlwind_trainables.count < 2
```
Comment (rpherbig) [inline validate.lic]: `setting` should be `settings` (note the `s`).

File: validate.lic [hunk 22]
```diff
@@ -277,6 +277,46 @@ class DRYamlValidator
...
+      if !settings.weapon_training.has_key?(trainable)
+        error("#{trainable} not found in weapon_training! All whirlwind_trainables need to also be listed in weapon_training.") 
+        return
+      end
+    }
+  end
+
+  def assert_whirlwind_trainables_are_not_swappables(settings)
+    return unless settings.whirlwind_trainables
+    swappables = ['war sword', 'bastard sword', 'throwing spike', 'bar mace', ' half-handled riste', 'riste', 'thick-bladed fan', 'boarding axe', 'broadaxe', 'splitting maul']
```
Comment (rpherbig) [inline validate.lic]: Are these all of the swappable templates?

You could instead take the string from `weapon_training` and look it up in the gear list, then check its `swappable` attribute.

File: validate.lic [hunk 23]
```diff
@@ -277,6 +277,46 @@ class DRYamlValidator
...
+
+    swappables.each { |desc|
+        if settings.weapon_training.has_value?(desc)
```
Comment (rpherbig) [inline validate.lic]: You could `swappables.select { |desc| settings.weapon_training.has_value?(desc) }.each { |desc| ... }` to make this code a bit cleaner.

File: validate.lic [hunk 24]
```diff
@@ -277,6 +277,46 @@ class DRYamlValidator
...
+          warn("#{desc} is a swappable! We don't support the swapping of weapon types for whirlwinding. We found this to be too cumbersome and take too much time!")
```
Comment (rpherbig) [inline validate.lic]: If this will break whirlwinding, I would expect these to be `error`s instead of `warning`s.

**Comment by rpherbig** [comment]
@jayRyan24 Please update elanthipedia with the setting information.
